source/slang/slang-ir-link.cpp

```diff
@@ -1519,6 +1520,138 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+        
+        auto std430 = builder._createInst(sizeof(IRTypeLayoutRules), builder.getType(kIROp_Std430BufferLayoutType), kIROp_Std430BufferLayoutType);
+        IRGLSLShaderStorageBufferType* storageBuffer;
+        {
+            IRInst* ops[] = { structType, std430 };
+            storageBuffer = builder.createGLSLShaderStorableBufferType(2, ops);
+        }
+
+        instToSwitch->setFullType(storageBuffer);
+        
+        // All references to a atomic_uint need to be an element ref. to emulate storage buffer usage
+        // All function calls must be inlined since storage buffers cannot pass as parameters to atomic methods
+        for (auto& i : bindingToInstList.second)
+        {
+            int64_t currOffset = int64_t(i->findDecoration<IRGLSLOffsetDecoration>()->getOffset()->getValue());
+
+            // we need a next node to be stored since the following code
+            // changes IRUse* of the use->user node, meaning we will lose
+            // our IRUse list of the atomic_uint being swapped out
+            IRUse* next = nullptr;
+            for (auto use = i->firstUse; use; use = next)
+            {
+                next = use->nextUse;
+                auto user = use->user;
+                
+                switch (user->getOp())
+                {
+                case kIROp_StructFieldLayoutAttr:
+                {
+                    // Definitions do nothing if unused
+                    break;
+                }
+                case kIROp_Call:
+                { 
+                    builder.setInsertBefore(user);
+                    auto fieldAddress = builder.emitFieldAddress(
+                        builder.getPtrType(_dataField->getFieldType()),
+                        instToSwitch,
+                        _dataField->getKey()
+                        );
+                    auto elementAddr = builder.emitElementAddress(
+                        builder.getPtrType(builder.getUIntType()),
+                        fieldAddress,
+                        builder.getIntValue(builder.getIntType(), currOffset/4));
+
+                    user->setOperand(1, elementAddr);
+                    auto funcTypeInst = (user->getOperand(0));
+                    auto funcType = funcTypeInst->getFullType();
+
+                    auto paramReplacment = builder.getInOutType(builder.getUIntType());
+                    funcType->getOperand(1)->replaceUsesWith(paramReplacment);
+                    builder.addForceInlineDecoration(funcTypeInst);
+
+                    break;
+                }
+                }
+            }
+            if (i->typeUse.usedValue->getOp() == kIROp_GLSLAtomicUintType)
+            {
+                i->removeAndDeallocate();
+            }
+        }
+    }
+}
+
+void GLSLTypeReplacmentProcessing(IRSpecContext* context, TargetProgram* targetProgram, IRModule* irModule)
+{
+    if (!targetProgram->getOptionSet().getBoolOption(CompilerOptionName::AllowGLSL)) return;
```

Comment:
There was a question about if we will have a chance to fix the coding style mistakes.
The answer is no.

It may sound contradicting to what I have said but a part of the reasons is to respect whoever wrote whatever code.

In practice, making cosmetic changes will cause big troubles when you need to trace back to where a mistake came from. I once removed a few white-lines and people kept asking me about how the function works for multiple years. I had to keep telling them to trace back deeper in the history.

Reply:
understood, that makes sense and why getting it correct initially is so important